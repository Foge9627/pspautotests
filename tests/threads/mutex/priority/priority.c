#include "../sub_shared.h"

SceUID mutex;

static int threadFunction(int argSize, void* argPointer) {
	int num = argPointer ? *((int*)argPointer) : 0;
	printf("A%d\n", num);
	sceKernelLockMutexCB(mutex, 1, NULL);
	printf("B%d\n", num);

	return 0;
}

void execPriorityTests(int attr) {
	SceUID threads[5];
	int test[5] = {1, 2, 3, 4, 5};

	mutex = sceKernelCreateMutex("mutex1", attr, 1, NULL);

	int i;
	for (i = 0; i < 5; i++) {
		threads[i] = CREATE_PRIORITY_THREAD(threadFunction, 0x16 - i);
		sceKernelStartThread(threads[i], sizeof(int), (void*)&test[i]);
	}

	// This one intentionally is an invalid unlock.
	sceKernelDelayThread(10 * 1000);
	printf("Unlocking...\n");
	printf("Unlocked 2? %08X\n", sceKernelUnlockMutex(mutex, 2));

	sceKernelDelayThread(10 * 1000);
	printf("Unlocking...\n");
	printf("Unlocked 1? %08X\n", sceKernelUnlockMutex(mutex, 1));

	sceKernelDelayThread(10 * 1000);
	printf("Delete: %08X\n", sceKernelDeleteMutex(mutex));
	printf("\n\n");
}

int main(int argc, char **argv) {
	execPriorityTests(0x000);
	execPriorityTests(0x100);
	return 0;
}