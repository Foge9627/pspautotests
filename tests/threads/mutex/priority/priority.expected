A1
A2
A3
A4
A5
Unlocking...
Unlocked 2? 800201BD
Unlocking...
B1
Unlocked 1? 00000000
B2
B3
B4
B5
Delete: 00000000


A1
A2
A3
A4
A5
Unlocking...
Unlocked 2? 800201BD
Unlocking...
B5
Unlocked 1? 00000000
B4
B3
B2
B1
Delete: 00000000


